File: src/util/dev/Idempotent.java
package util.dev;

import java.lang.annotation.*;

/**
 * Denotes whether operation is idempotent, i.e., invoking it multiple times has the same effect as invoking it once, or
 * in other words, any subsequent invoke has no (side) effect at all.
 */
@Documented
@Target(ElementType.METHOD)
@Retention(RetentionPolicy.RUNTIME)
public @interface Idempotent {
	/** Default true. */
	boolean value() default true;
}

File: src/util/dev/Idempotent.kt
package util.dev

/**
 * Denotes whether operation (with a side effect) is idempotent - invoking it subsequently has the same effect as
 * invoking it once, i.e., any subsequent invokes have no (side) effect at all.
 */
@MustBeDocumented
@Retention(AnnotationRetention.SOURCE)
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
annotation class Idempotent